### DISCUSSION TITLE
Initialize Isolate spawned from native code

### DISCUSSION THREAD

**Discussion by Geparder** [cofu-app/cbl-dart#393] [General]
How to initialize in isolate?

**Comment by blaugold** [comment]
Thanks for this question!

1. Ensure that you have initialized the primary isolate with either `CouchbaseLiteDart.init` or `CouchbaseLiteFlutter.init`.
2. Send `CouchbaseLite.context` to the secondary isolate.
3. In the secondary isolate, call `CouchbaseLite.initSecondary` with the context.

The example below demonstrates all the steps:

```dart
// ignore_for_file: avoid_print

import 'dart:isolate';

import 'package:cbl/cbl.dart';
import 'package:cbl_dart/cbl_dart.dart';

Future<void> main() async {
  // Couchbase Lite needs to be initialized before it can be used.
  await CouchbaseLiteDart.init(edition: Edition.community);

  final onExit = ReceivePort();
  await Isolate.spawn(
    _secondaryMain,
    CouchbaseLite.context,
    onExit: onExit.sendPort,
  );
  await onExit.first;
}

Future<void> _secondaryMain(Object context) async {
  await CouchbaseLite.initSecondary(context);

  final db = await Database.openAsync('chat-messages');

  final doc = MutableDocument({
    'type': 'message',
    'body': 'Heyo',
    'from': 'Alice',
  });
  print(doc);

  await db.saveDocument(doc);
  print(doc);

  await db.close();
}
```

**Comment by blaugold** [comment]
@Geparder  Not sure what happened to your last comment:

> Thank you!
The isolate is run from native Java code (in a service). I need to pass the database context to the service from flutter.
This I can only be done in the form of string (serialize).
But the IsolateContext class is private. And it cannot be deserialized to save it in the service.
>
> ...I'm sorry for my bad english

This use case is not currently supported. I have opened https://github.com/cbl-dart/cbl-dart/issues/394 to track implementing support for it.
